### PR TITLE
Align to changes in oM - moving Reflection_oM objects to Base_oM -> u…pdating usings and codings to the new object locations

### LINKED ISSUE #1645: Global: Try to remove the dependency between the BHoM engine and the Reflection Engine
Migrate the method in the reflections that are used BHoM-Engine to the BHoM engine itself. 
Also migrate most of reflection oM to the base.

### LINKED ISSUE #2671: Need to reorganise the code between Reflection Engine and Base Engine ?
Right now, the Base Engine depends on the Reflection Engine. Which means that the reflection Engine cannot leverage what is available in the Base Engine. Most of the time it is fine since the Base Engine doesn't contain much anyways but the new `Enumeration` objects for example need to be handled properly in the `SetProperty` method, which requires the access to the `Create.Enumeration` method from the Base engine (See this PR: https://github.com/BHoM/BHoM_Engine/pull/2637 where I had to duplicate the code to avoid circular dependencies).

I think we also discussed before the possibility to migrate the `RecordError/RecordWarning` methods from the reflection Engine to the base Engine. A few other methods that are specifically targeted toward the BHoM objects would probably make sense to migrate.

Here's the list of Reflection methods that are used in the Base Engine right now:
- RecordError / RecordWarning
- BHoMTypeList
- ExtensionMethodToCall and RunExtensionMethod
- Personally, I don't have an issue with them being part of the Base Engine instead as they are central to the functioning of the BHoM.

The attributes (input and output attributes) are also in reflection and are used in the Base Engine. I wouldn't mind seing those ones migrated as well.

### PR DESCRIPTION
**Pull request by FraserGreenroyd**
### NOTE: Depends on 
https://github.com/BHoM/BHoM/pull/1339

   
### Issues addressed by this PR
Fixes #1645
Fixes #2671 

### Changelog
 - Removes Reflection_oM dependencies from all projects
 - Updates all usings from BH.oM.Reflection -> BH.oM.Base
 - Moves methods from Reflection_Engine to Base_Engine in line with previous agreements in issues 1645 and 2671

### Additional comments
Please read associated issues by the linked PR for additional background.

I have moved all methods from Reflection_Engine to BHoM_Engine where BHoM_Engine required them. If BHoM_Engine did not require them, then they have stayed in Reflection_Engine for now.

The attribute query files have stayed in Reflection_Engine for now as I feel they are appropriate there as this is code asking itself for its documentation. But am happy to discuss.

Some instances of `Deprecated` attribute have been changed and versioned and then the methods removed, others have been replaced with a `ToBeRemoved` attribute for future work - this is primarily in the Structure_Engine which has a lot of `Deprecated` methods that may impact other toolkits if removed now, and quickly fall out of scope of the original work of this PR.

@adecler if you could review the engine over night, @IsakNaslundBh plans to review things tomorrow with an aim to merge in the installer repos ASAP to avoid holding up other milestone development, as everything is now on hold pending these being merged, so keen to achieve quick resolution and possibly break further items into new tasks if necessary. Obviously anything major missed will need to be actioned, but if it's small and could be done in a new, smaller scoped PR (now the dependency on reflection_engine/oM is broken for many toolkits) that would probably be ideal.

### CODE REVIEW

File: Structure_Engine/Create/Loads/LoadCombination.cs
Comment (IsakNaslundBh) [inline Structure_Engine/Create/Loads/LoadCombination.cs:60]: Generally looks fine. details: Could get rid of the ToList() the factoredCases as it already is a list, aparat from that looks good.

Comment (FraserGreenroyd) [reply Structure_Engine/Create/Loads/LoadCombination.cs]: Resolved in [this commit](https://github.com/BHoM/BHoM_Engine/pull/2729/commits/4e57bc6a6e7af5e6fb404a5d605ebc2e659d6ce1)

**Comment by FraserGreenroyd** [comment]
Lots of files changed there... @IsakNaslundBh some specifics on Structure_Engine as discussed, contained in a couple commits to try and make it easier for you.

@adecler can you check you're happy with what's been moved from Reflection_Engine -> Base_Engine and double check if I've missed anything? I've tried checking against #1645 and #2671 but a second pair of eyes would be good on that.

I've updated the PR body with some more information too based on this.

I'll now be updating the dependant toolkits down the line - so further changes here will be good to capture ASAP if needed. Cheers!

**Review by adecler — COMMENTED** [review]
There's a few extra undocumented micro changes that are difficult to spot (e.g. change of method name or multiple files now grouped into one) so it is difficult not to miss anything in the review.

There is obviously a bit of a blurry line between the public Reflection methods that have to be inside the Base Engine because they are used by other base methods and the ones that stayed in the reflection engine but I guess that cannot be avoided. E.g.: `Base.Query.GetAllPropertyFullNames` vs `Reflection.Query.PropertyNames`. If the attribute methods gets migrated as well, I wonder if there is going to be a point where the reflection engine is too small and too similar to base methods to make sense on it's own. But a problem for another time 😄 .

@al-fisher, I wonder what your opinion is about `IsPrototype` staying in the Reflection engine ?

Other than that, it looks good to me 👍

File: .ci/code/Engine_Test/Engine_Test.csproj
Comment (FraserGreenroyd) [reply .ci/code/Engine_Test/Engine_Test.csproj:39]: 👍 all the repos will be getting project compliance run on them as a few will have encountered this problem I suspect across the work done.

File: BHoM_Engine/Compute/Record.cs
Comment (FraserGreenroyd) [reply BHoM_Engine/Compute/Record.cs:84]: Am happy to split them back out into 4 files if desired.

File: BHoM_Engine/Compute/LoadAssembly.cs
Comment (IsakNaslundBh) [reply BHoM_Engine/Compute/LoadAssembly.cs:66]: Understand after chat on teams with @FraserGreenroyd that the name change was done to not use the Reflect word in the base engine. Can I suggest we then instead use Extract and call this method `ExtractAssembly`. That is after all what all the private methods used by this method is called.

File: BHoM_Engine/Compute/LoadNonBHoMTypes.cs
Comment (IsakNaslundBh) [reply BHoM_Engine/Compute/LoadNonBHoMTypes.cs:43]: as above

**Comment by FraserGreenroyd** [comment]
> There is obviously a bit of a blurry line between the public Reflection methods that have to be inside the Base Engine because they are used by other base methods and the ones that stayed in the reflection engine but I guess that cannot be avoided. E.g.: `Base.Query.GetAllPropertyFullNames` vs `Reflection.Query.PropertyNames`. If the attribute methods gets migrated as well, I wonder if there is going to be a point where the reflection engine is too small and too similar to base methods to make sense on it's own. But a problem for another time 😄 .

Yeah the current work has been done to break the dependency for BHoM_Engine on Reflection_Engine, and thus break the dependency on every toolkit requiring Reflection_Engine to only a handful. I am envisaging we will do another phase of work moving some other methods over as well in smaller surgical PRs in the future, as there is still a lot in Reflection_Engine that is not strictly pure reflection.
 
> @al-fisher, I wonder what your opinion is about `IsPrototype` staying in the Reflection engine ?

This will move in a second wave 😄 

> Other than that, it looks good to me 👍

🥳

**Comment by al-fisher** [comment]
>> @al-fisher, I wonder what your opinion is about IsPrototype staying in the Reflection engine ?

>This will move in a second wave 😄

👍

File: BHoM_Engine/Compute/LoadAssembly.cs
Comment (FraserGreenroyd) [reply BHoM_Engine/Compute/LoadAssembly.cs]: Have renamed to `Extract`.

File: BHoM_Engine/Compute/LoadNonBHoMTypes.cs
Comment (FraserGreenroyd) [reply BHoM_Engine/Compute/LoadNonBHoMTypes.cs]: Have renamed

**Comment by pawelbaran** [comment]
@FraserGreenroyd I was following the notifications and noticed that you renamed `LoadAssembly` to `ExtractAssembly` - as much as I agree with it, I thought `ReflectAssembly` could be even more precise?

**Comment by FraserGreenroyd** [comment]
> @FraserGreenroyd I was following the notifications and noticed that you renamed `LoadAssembly` to `ExtractAssembly` - as much as I agree with it, I thought `ReflectAssembly` could be even more precise?

Following the discussions on the use for Reflection_Engine vs Base_Engine, I am against anything in the Base_Engine `Reflecting` anything, so I wanted to break that link. I reviewed the code and while it could be argued that the method is performing reflection in so much as we are asking the assembly about its types, we're not necessarily asking the assembly to do something with the answer to the question. We're more extracting (or loading as I first went with) the types from the assembly into our eco-system for use later by other processes. Ended up on `extract` following discussion with @IsakNaslundBh but am happy to do more discussion on another name, but I am keen on keeping `reflection` out of Base_Engine if we can 😄

**Comment by pawelbaran** [comment]
> Following the discussions on the use for Reflection_Engine vs Base_Engine, I am against anything in the Base_Engine `Reflecting` anything, so I wanted to break that link. I reviewed the code and while it could be argued that the method is performing reflection in so much as we are asking the assembly about its types, we're not necessarily asking the assembly to do something with the answer to the question. We're more extracting (or loading as I first went with) the types from the assembly into our eco-system for use later by other processes. Ended up on `extract` following discussion with @IsakNaslundBh but am happy to do more discussion on another name, but I am keen on keeping `reflection` out of Base_Engine if we can 😄

No worries, no strong opinion here (as usual with the names and colours 😃)

**Review by IsakNaslundBh — DISMISSED** [review]
Changes with regards to migration looks good to me, and for any strangling bits to be ironed out in smaller PRs as mentioned by @FraserGreenroyd .

Have reviewed the structure_Engine as well, as it was done in a separate commit, and all changes there looks good.

Approving and letting BHoMBot catch if things are not compiling.

**Comment by adecler** [comment]
The Reflection engine is still referenced by all the other engines. 
There are also 58 `using BH.Engine.Reflection;` thatare now irrelevant.

**Comment by FraserGreenroyd** [comment]
@BHoMBot check installer

**Comment by FraserGreenroyd** [comment]
@BHoMBot this is a DevOps instruction. I am requesting neutral checks on: core, code-compliance, copyright-compliance, dataset-compliance, documentation-compliance, null-handling, project-compliance, ready-to-merge, serialisation

**Review by jamesramsden-bh — APPROVED** [review]
Approving based on discussion with @FraserGreenroyd as part of the merge party to merge in this work. Core work was reviewed by @IsakNaslundBh and @adecler during the week. This approval is given to allow merge, with checks on installer and versioning from BHoMBot being sufficient to show compilation upgrade to the core changes reviewed by others.

**Pull request merged**
